File: device/cuda/activation/native/standard.c
#include "standard.h"
#include "activation.h"
#include "../internal/bridge/core_private.h"

#include <stdint.h>
#include <string.h>

static const char* cuda_standard_unary_operation_name(int operation) {
    switch (operation) {
    case CUDAUnaryFloat32Relu:
        return "relu";
    case CUDAUnaryFloat32Exp:
        return "exp";
    case CUDAUnaryFloat32Log:
        return "log";
    case CUDAUnaryFloat32Tanh:
        return "tanh";
    case CUDAUnaryFloat32Gelu:
        return "gelu";
    case CUDAUnaryFloat32Sigmoid:
        return "sigmoid";
    case CUDAUnaryFloat32Silu:
        return "silu";
    case CUDAUnaryFloat32Swish:
        return "swish";
    case CUDAUnaryFloat32Softsign:
        return "softsign";
    case CUDAUnaryFloat32ELU:
        return "elu";
    case CUDAUnaryFloat32SELU:
        return "selu";
    case CUDAUnaryFloat32LeakyReLU:
        return "leaky_relu";
    case CUDAUnaryFloat32HardSigmoid:
        return "hardsigmoid";
    case CUDAUnaryFloat32HardSwish:
        return "hardswish";
    case CUDAUnaryFloat32Log1p:
        return "log1p";
    case CUDAUnaryFloat32Expm1:
        return "expm1";
    case CUDAUnaryFloat32CELU:
        return "celu";
    case CUDAUnaryFloat32Softplus:
        return "softplus";
    case CUDAUnaryFloat32Mish:
        return "mish";
    case CUDAUnaryFloat32LogSigmoid:
        return "log_sigmoid";
    case CUDAUnaryFloat32GeluTanh:
        return "gelu_tanh";
    case CUDAUnaryFloat32HardTanh:
        return "hardtanh";
    case CUDAUnaryFloat32HardGelu:
        return "hard_gelu";
    case CUDAUnaryFloat32QuickGelu:
        return "quick_gelu";
    case CUDAUnaryFloat32TanhShrink:
        return "tanh_shrink";
    default:
        return NULL;
    }
}

int cuda_dispatch_unary_elementwise(
    CUDADeviceRef contextRef,
    int operation,
    int elementDType,
    CUDABufferRef inputRef,
    CUDABufferRef outRef,
    uint32_t count,
    uint64_t completionToken,
    CUDAStatus* status
) {
    cuda_activation_status_clear(status);

    if (count == 0) {
        return 0;
    }

    if (inputRef == NULL || outRef == NULL) {
        cuda_activation_status_set(status, -2, "nil CUDA buffer");
        return -2;
    }

    const char* operationName = cuda_standard_unary_operation_name(operation);

    if (operationName == NULL) {
        cuda_activation_status_set(status, -6, "unknown CUDA standard unary operation");
        return -6;
    }

    const char* dtypeSuffix = cuda_activation_element_dtype_suffix(elementDType);

    if (dtypeSuffix == NULL) {
        cuda_activation_status_set(status, -6, "unknown CUDA standard unary dtype");
        return -6;
    }

    char kernelName[128];
    int nameCode = cuda_activation_compose_kernel_name(
        kernelName,
        sizeof(kernelName),
        operationName,
        dtypeSuffix,
        status
    );

    if (nameCode != 0) {
        return nameCode;
    }

    const char* moduleSource = cuda_activation_module_source();

    if (moduleSource == NULL) {
        cuda_activation_status_set(status, -7, "CUDA activation module source not registered");
        return -7;
    }

    CUDAContext* context = NULL;
    CUDAStreamRef stream = NULL;
    int prepareCode = cuda_context_prepare(contextRef, status, &context, &stream);

    if (prepareCode != 0) {
        return prepareCode;
    }

    CUDAKernelRef kernel = cuda_get_kernel(context, moduleSource, kernelName, status);

    if (kernel == NULL) {
        return status != NULL && status->code != 0 ? status->code : -7;
    }

    void* inputPtr = cuda_buffer_device_ptr(inputRef);
    void* outputPtr = cuda_buffer_device_ptr(outRef);
    void* args[] = {&inputPtr, &outputPtr, &count};
    int launchCode = cuda_launch_1d(context, kernel, stream, count, args, sizeof(args), status);

    if (launchCode != 0) {
        return launchCode;
    }

    cuda_track_completion(context, stream, completionToken, NULL, status);
    return 0;
}

File: device/metal/activation/native/standard.m
#include "standard.h"
#include "activation.h"
#include "../internal/bridge/core_private.h"

#include <Foundation/Foundation.h>
#include <Metal/Metal.h>

static const char* metal_standard_unary_operation_name(int operation) {
    switch (operation) {
    case MetalUnaryFloat32Relu: return "relu";
    case MetalUnaryFloat32Abs: return "abs";
    case MetalUnaryFloat32Neg: return "neg";
    case MetalUnaryFloat32Square: return "square";
    case MetalUnaryFloat32Recip: return "recip";
    case MetalUnaryFloat32Sqrt: return "sqrt";
    case MetalUnaryFloat32Sign: return "sign";
    case MetalUnaryFloat32Rsqrt: return "rsqrt";
    case MetalUnaryFloat32Exp: return "exp";
    case MetalUnaryFloat32Log: return "log";
    case MetalUnaryFloat32Sin: return "sin";
    case MetalUnaryFloat32Cos: return "cos";
    case MetalUnaryFloat32Tanh: return "tanh";
    case MetalUnaryFloat32Gelu: return "gelu";
    case MetalUnaryFloat32Sigmoid: return "sigmoid";
    case MetalUnaryFloat32Silu: return "silu";
    case MetalUnaryFloat32Swish: return "swish";
    case MetalUnaryFloat32Softsign: return "softsign";
    case MetalUnaryFloat32ELU: return "elu";
    case MetalUnaryFloat32SELU: return "selu";
    case MetalUnaryFloat32LeakyReLU: return "leaky_relu";
    case MetalUnaryFloat32HardSigmoid: return "hardsigmoid";
    case MetalUnaryFloat32HardSwish: return "hardswish";
    case MetalUnaryFloat32Log1p: return "log1p";
    case MetalUnaryFloat32Expm1: return "expm1";
    case MetalUnaryFloat32CELU: return "celu";
    case MetalUnaryFloat32Softplus: return "softplus";
    case MetalUnaryFloat32Mish: return "mish";
    case MetalUnaryFloat32LogSigmoid: return "log_sigmoid";
    case MetalUnaryFloat32GeluTanh: return "gelu_tanh";
    case MetalUnaryFloat32HardTanh: return "hardtanh";
    case MetalUnaryFloat32HardGelu: return "hard_gelu";
    case MetalUnaryFloat32QuickGelu: return "quick_gelu";
    case MetalUnaryFloat32TanhShrink: return "tanh_shrink";
    default: return NULL;
    }
}

static int metal_standard_prepare(
    MetalContext* context,
    const char* kernelName,
    MetalStatus* status,
    id<MTLComputePipelineState>* pipeline
) {
    if (context == NULL || context->queue == NULL) {
        metal_activation_status_set(status, -1, "invalid Metal context");
        return -1;
    }

    *pipeline = metal_get_pipeline(context, kernelName, status);

    if (*pipeline == nil) {
        return status != NULL && status->code != 0 ? status->code : -7;
    }

    return 0;
}

int metal_dispatch_unary_elementwise(
    MetalDeviceRef contextRef,
    int operation,
    int elementDType,
    MetalBufferRef inputRef,
    MetalBufferRef outRef,
    uint32_t count,
    uint64_t completionToken,
    MetalStatus* status
) {
    @autoreleasepool {
        metal_activation_status_clear(status);

        if (count == 0) {
            return 0;
        }

        if (inputRef == NULL || outRef == NULL) {
            metal_activation_status_set(status, -2, "nil Metal buffer");
            return -2;
        }

        const char* operationName = metal_standard_unary_operation_name(operation);

        if (operationName == NULL) {
            metal_activation_status_set(status, -6, "unknown Metal standard unary operation");
            return -6;
        }

        const char* dtypeSuffix = metal_activation_element_dtype_suffix(elementDType);

        if (dtypeSuffix == NULL) {
            metal_activation_status_set(status, -6, "unknown Metal standard unary dtype");
            return -6;
        }

        char kernelName[128];
        int nameCode = metal_activation_compose_kernel_name(
            kernelName,
            sizeof(kernelName),
            operationName,
            dtypeSuffix,
            status
        );

        if (nameCode != 0) {
            return nameCode;
        }

        id<MTLComputePipelineState> pipeline = nil;
        int prepareCode = metal_standard_prepare(
            (MetalContext*)contextRef,
            kernelName,
            status,
            &pipeline
        );

        if (prepareCode != 0) {
            return prepareCode;
        }

        id<MTLCommandBuffer> commandBuffer;
        id<MTLComputeCommandEncoder> encoder = metal_get_encoder((MetalContext*)contextRef, &commandBuffer);

        [encoder setComputePipelineState:pipeline];
        [encoder setBuffer:(__bridge id<MTLBuffer>)inputRef offset:0 atIndex:0];
        [encoder setBuffer:(__bridge id<MTLBuffer>)outRef offset:0 atIndex:1];
        [encoder setBytes:&count length:sizeof(count) atIndex:2];

        NSUInteger threadWidth = [pipeline threadExecutionWidth];

        if (threadWidth == 0) {
            threadWidth = 1;
        }

        NSUInteger vectorCount = (NSUInteger)((count + 3) / 4);
        [encoder
            dispatchThreads:MTLSizeMake(vectorCount, 1, 1)
            threadsPerThreadgroup:MTLSizeMake(threadWidth, 1, 1)
        ];
        metal_track_command_completion((MetalContext*)contextRef, commandBuffer, completionToken, NULL);
        metal_end_encoder((MetalContext*)contextRef, encoder, commandBuffer);

        return 0;
    }
}
